### PR TITLE
Analyzer - apply limit and offset settings - second attempt

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
not for changelog - part of #42648

### CODE REVIEW

File: src/Planner/Planner.cpp [hunk 1]
```diff
@@ -216,12 +216,17 @@ class QueryAnalysisResult
             limit_length = query_node.getLimit()->as<ConstantNode &>().getValue().safeGet<UInt64>();
         }
 
+        if (settings.limit)
```
Comment (kitaisreal) [inline src/Planner/Planner.cpp:219]: This will not work properly in case of distributed queries. Because it relies that we will send context with specific settings, and it does not look reliable.

File: src/Analyzer/QueryTreeBuilder.cpp [hunk 2]
```diff
@@ -233,6 +233,17 @@ QueryTreeNodePtr QueryTreeBuilder::buildSelectExpression(const ASTPtr & select_q
     auto select_settings = select_query_typed.settings();
     SettingsChanges settings_changes;
 
+    if (is_subquery)
```
Comment (kitaisreal) [inline src/Analyzer/QueryTreeBuilder.cpp:236]: This is not right. Because `buildSelectExpression` can be called for subquery but without is_subquery, for example for `view(SELECT 1)`, `SELECT 1` is not subquery.
I suspect that we need to add local workaround in this place, and do not handle it anywhere else. This `limit`, `offset` settings does not look easily understandable when we send them to remote servers.
Here we need to save `limit`, `offset` settings, and remove them from our our updated context.
Then we can get minimum between this `limit` and LIMIT in query if it exists using `if` function. If LIMIT does not exists we can just add constant node with `limit` setting instead of LIMIT.
For `offset`, we must use `plus` function with OFFSET in query if it exists. If OFFSET does not exists we can just add constant node with `offset` setting instead of OFFSET.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Analyzer/QueryTreeBuilder.cpp (hunk 2 above)
Comment (yakov-olkhovskiy) [reply src/Analyzer/QueryTreeBuilder.cpp:236]: `view` argument receives its context with `getViewContext` function - I fixed it there - I added test for `view` and it works

File: src/Analyzer/QueryTreeBuilder.cpp [hunk 3]
```diff
@@ -233,22 +233,43 @@ QueryTreeNodePtr QueryTreeBuilder::buildSelectExpression(const ASTPtr & select_q
...
     SettingsChanges settings_changes;
 
-    if (is_subquery)
+    /// We are going to remove settings LIMIT and OFFSET and
+    /// further replace them with corresponding expression nodes
+    UInt64 limit = 0;
+    UInt64 offset = 0;
+
+    /// remove global settings limit and offset
+    if (const Settings & settings_ref = updated_context->getSettingsRef(); settings_ref.limit || settings_ref.offset)
```
Comment (kitaisreal) [inline src/Analyzer/QueryTreeBuilder.cpp:242]: Consider to rename `settings_ref`, into just `const auto & settings =`.

File: src/Analyzer/QueryTreeBuilder.cpp [hunk 4]
```diff
@@ -323,12 +355,32 @@ QueryTreeNodePtr QueryTreeBuilder::buildSelectExpression(const ASTPtr & select_q
     if (select_limit_by)
         current_query_tree->getLimitByNode() = buildExpressionList(select_limit_by, current_context);
 
+    /// combine limit expression with limit setting
```
Comment (kitaisreal) [inline src/Analyzer/QueryTreeBuilder.cpp:358]: Comment should be started with capital letter.

File: src/Analyzer/QueryTreeBuilder.cpp [hunk 5]
```diff
@@ -323,12 +355,32 @@ QueryTreeNodePtr QueryTreeBuilder::buildSelectExpression(const ASTPtr & select_q
...
+        function_node->getArguments().getNodes().push_back(buildExpression(select_limit, current_context));
+        function_node->getArguments().getNodes().push_back(std::make_shared<ConstantNode>(limit));
+        current_query_tree->getLimit() = function_node;
+    }
+    else if (limit)
+        current_query_tree->getLimit() = std::make_shared<ConstantNode>(limit);
+    else if (select_limit)
         current_query_tree->getLimit() = buildExpression(select_limit, current_context);
 
+    /// combine offset expression with offset setting
```
Comment (kitaisreal) [inline src/Analyzer/QueryTreeBuilder.cpp:372]: Comment should be started with capital letter.

File: src/Analyzer/QueryTreeBuilder.cpp [hunk 6]
```diff
@@ -323,12 +355,32 @@ QueryTreeNodePtr QueryTreeBuilder::buildSelectExpression(const ASTPtr & select_q
...
 
+    /// combine limit expression with limit setting
     auto select_limit = select_query_typed.limitLength();
-    if (select_limit)
+    if (select_limit && limit)
+    {
+        auto function_node = std::make_shared<FunctionNode>("least");
+        function_node->getArguments().getNodes().push_back(buildExpression(select_limit, current_context));
+        function_node->getArguments().getNodes().push_back(std::make_shared<ConstantNode>(limit));
+        current_query_tree->getLimit() = function_node;
```
Comment (kitaisreal) [inline src/Analyzer/QueryTreeBuilder.cpp:365]: Consider to use `std::move(function_node)`.

File: src/Analyzer/QueryTreeBuilder.cpp [hunk 7]
```diff
@@ -323,12 +355,32 @@ QueryTreeNodePtr QueryTreeBuilder::buildSelectExpression(const ASTPtr & select_q
...
 
+    /// combine offset expression with offset setting
     auto select_offset = select_query_typed.limitOffset();
-    if (select_offset)
+    if (select_offset && offset)
+    {
+        auto function_node = std::make_shared<FunctionNode>("plus");
+        function_node->getArguments().getNodes().push_back(buildExpression(select_offset, current_context));
+        function_node->getArguments().getNodes().push_back(std::make_shared<ConstantNode>(offset));
+        current_query_tree->getOffset() = function_node;
```
Comment (kitaisreal) [inline src/Analyzer/QueryTreeBuilder.cpp:379]: Consider to use `std::move(function_node)`.

File: src/Common/SettingsChanges.cpp [hunk 8]
```diff
@@ -46,4 +46,29 @@ Field * SettingsChanges::tryGet(std::string_view name)
     return &change->value;
 }
 
+bool SettingsChanges::insertSetting(std::string_view name, const Field & value)
+{
+    if (std::find_if(begin(), end(), [&name](const SettingChange & change) { return change.name == name; }) != end())
```
Comment (kitaisreal) [inline src/Common/SettingsChanges.cpp:51]: Cosider to extract iterator out of if statement.

Comment (kitaisreal) [inline src/Common/SettingsChanges.cpp:51]: We should just copy string view and do not take reference in lambda `[&name]`.

File: src/Common/SettingsChanges.cpp [hunk 9]
```diff
@@ -46,4 +46,29 @@ Field * SettingsChanges::tryGet(std::string_view name)
...
+        return false;
+    emplace_back(name, value);
+    return true;
+}
+
+void SettingsChanges::setSetting(std::string_view name, const Field & value)
+{
+    if (auto * v = tryGet(name))
```
Comment (kitaisreal) [inline src/Common/SettingsChanges.cpp:59]: Consider rename `v` variable into something more informative. For example `setting_value`.

File: src/Common/SettingsChanges.cpp [hunk 10]
```diff
@@ -46,4 +46,29 @@ Field * SettingsChanges::tryGet(std::string_view name)
...
+        *v = value;
+    else
+        insertSetting(name, value);
+}
+
+bool SettingsChanges::removeSetting(std::string_view name)
+{
+    auto it = std::find_if(begin(), end(), [&name](const SettingChange & change) { return change.name == name; });
```
Comment (kitaisreal) [inline src/Common/SettingsChanges.cpp:67]: The same as in `insertSetting` method.

File: src/Common/SettingsChanges.h [hunk 11]
```diff
@@ -28,6 +28,13 @@ class SettingsChanges : public std::vector<SettingChange>
     bool tryGet(std::string_view name, Field & out_value) const;
     const Field * tryGet(std::string_view name) const;
     Field * tryGet(std::string_view name);
+
+    /// inserts element if doesn't exists and returns true, else just returns false
```
Comment (kitaisreal) [inline src/Common/SettingsChanges.h:32]: Comments should start with capital letter.

Comment (kitaisreal) [inline src/Common/SettingsChanges.h:32]: Consider to rewrite `else just returns false` into `otherwise returns false`.

File: src/Common/SettingsChanges.h [hunk 12]
```diff
@@ -28,6 +28,13 @@ class SettingsChanges : public std::vector<SettingChange>
...
+    bool insertSetting(std::string_view name, const Field & value);
+    /// sets element to value, inserts if doesn't exist
+    void setSetting(std::string_view name, const Field & value);
+    /// if element exists - removes it and returns true, else returns false
```
Comment (kitaisreal) [inline src/Common/SettingsChanges.h:36]: Consider to rewrite else just returns false into otherwise returns false.

File: src/Planner/Planner.cpp [hunk 13]
```diff
@@ -720,7 +720,7 @@ bool addPreliminaryLimitOptimizationStepIfNeeded(QueryPlan & query_plan,
 
     bool apply_limit = query_processing_info.getToStage() != QueryProcessingStage::WithMergeableStateAfterAggregation;
     bool apply_prelimit = apply_limit &&
-        query_node.hasLimit() &&
+        query_analysis_result.limit_length &&
```
Comment (kitaisreal) [inline src/Planner/Planner.cpp:723]: Unrelated changes in this file.

File: tests/queries/0_stateless/02667_analyzer_limit_settings.sql [hunk 14]
```diff
@@ -0,0 +1,30 @@
+SET allow_experimental_analyzer = 1;
+
+-- { echoOn }
+SET limit = 0;
+
+SELECT * FROM numbers(10);
```
Comment (kitaisreal) [inline tests/queries/0_stateless/02667_analyzer_limit_settings.sql:6]: Consider add test with distributed.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Analyzer/QueryTreeBuilder.cpp (hunk 3 above)
Comment (yakov-olkhovskiy) [reply src/Analyzer/QueryTreeBuilder.cpp:242]: it's only used to check if limit or offset present and further we use `settings` name for copy

File: src/Common/SettingsChanges.cpp (hunk 8 above)
Comment (yakov-olkhovskiy) [reply src/Common/SettingsChanges.cpp:51]: it's how it used in existing code above and I don't see it's anyhow important here

File: src/Common/SettingsChanges.h (hunk 11 above)
Comment (yakov-olkhovskiy) [reply src/Common/SettingsChanges.h:32]: "else just" is used to emphasize that existing value is not changed - "otherwise" would be ambiguous in this regard

File: src/Planner/Planner.cpp (hunk 13 above)
Comment (yakov-olkhovskiy) [reply src/Planner/Planner.cpp:723]: it's a bug fix - limit should be taken here from analysis result

Comment (kitaisreal) [reply src/Planner/Planner.cpp:723]: Where was bug ?
```
if (query_node.hasLimit())
{
    /// Constness of limit is validated during query analysis stage
    limit_length = query_node.getLimit()->as<ConstantNode &>().getValue().safeGet<UInt64>();
}
```
We should not use this if we want to check if query node has limit.

Comment (yakov-olkhovskiy) [reply src/Planner/Planner.cpp:723]: the use of `query_node.hasLimit()` is a bug - further we a calling a`ddPreliminaryLimitStep` passing `query_analysis_result` and using `query_analysis_result.limit_length`:
```
void addPreliminaryLimitStep(QueryPlan & query_plan,
    const QueryAnalysisResult & query_analysis_result,
    const PlannerContextPtr & planner_context,
    bool do_not_skip_offset)
{
    UInt64 limit_offset = query_analysis_result.limit_offset;
    UInt64 limit_length = query_analysis_result.limit_length;
...
```
so, apparently we should use query_analysis_result.limit_length there

File: src/Common/SettingsChanges.h [hunk 15]
```diff
@@ -28,6 +28,13 @@ class SettingsChanges : public std::vector<SettingChange>
     bool tryGet(std::string_view name, Field & out_value) const;
     const Field * tryGet(std::string_view name) const;
     Field * tryGet(std::string_view name);
+
+    /// Inserts element if doesn't exists and returns true, otherwise  just returns false
```
Comment (kitaisreal) [inline src/Common/SettingsChanges.h:32]: Double spaces `otherwise  just` between `otherwise` and `just`.

File: src/Common/SettingsChanges.h [hunk 16]
```diff
@@ -28,6 +28,13 @@ class SettingsChanges : public std::vector<SettingChange>
...
+    bool insertSetting(std::string_view name, const Field & value);
+    /// Sets element to value, inserts if doesn't exist
+    void setSetting(std::string_view name, const Field & value);
+    /// If element exists - removes it and returns true, otherwise  returns false
```
Comment (kitaisreal) [inline src/Common/SettingsChanges.h:36]: Double spaces `otherwise  returns` between `otherwise` and `returns`.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

**Review by kitaisreal — APPROVED** [review]
